### PR TITLE
[Pytorch-iOS] remove ONNX & Turn on `NO_API` for mobile build

### PR DESCRIPTION
**Pull request by xta0**
### Summary 
The iOS build was broken after this PR 👉 [23195](https://github.com/pytorch/pytorch/pull/23195/files) was merged, as there are two files still have dependency on ONNX. 
- `test.cpp` in `test/cpp/jit`
-  `export.cpp` in `torch/csrc/jit`

This PR is to remove ONNX completely from mobile build. 

### Test plan

- The `build_ios.sh` finished successfully.
- The `libtorch.a` can  be compiled and run on iOS devices

### CODE REVIEW

File: caffe2/CMakeLists.txt [hunk 1]
```diff
@@ -460,6 +460,11 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     ${TORCH_ROOT}/test/cpp/jit/test.cpp
     )
 
+  if (INTERN_BUILD_MOBILE)
+    list(REMOVE_ITEM TORCH_SRCS ${TORCH_SRC_DIR}/csrc/jit/export.cpp)
```
Comment (dzhulgakov) [inline caffe2/CMakeLists.txt]: REMOVE_ITEM is a bit magical, can we instead just add it only if the opposite condition is true?

**Review by dzhulgakov — CHANGES_REQUESTED** [review]

File: caffe2/CMakeLists.txt (hunk 1 above)
Comment (xta0) [reply caffe2/CMakeLists.txt]: You mean something like this?  
```
  if (NOT INTERN_BUILD_MOBILE)
    list(APPEND TORCH_SRCS ${TORCH_SRC_DIR}/csrc/jit/export.cpp)
    list(APPEND TORCH_SRCS ${TORCH_ROOT}/test/cpp/jit/test.cpp)
  endif()
```

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -457,13 +456,18 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
...
 
+  if (NOT INTERN_BUILD_MOBILE)
+    list(APPEND TORCH_SRCS ${TORCH_SRC_DIR}/csrc/jit/export.cpp)
+    list(APPEND TORCH_SRCS ${TORCH_ROOT}/test/cpp/jit/test.cpp)
+    if(NOT WIN32)
+      list(APPEND TORCH_SRCS ${TORCH_SRC_DIR}/csrc/jit/fuser/cpu/fused_kernel.cpp)
+    endif()
+  endif()
+
   if (NOT WIN32 AND NOT INTERN_BUILD_MOBILE)
```
Comment (ljk53) [inline caffe2/CMakeLists.txt]: delete the empty block?

File: caffe2/CMakeLists.txt [hunk 3]
```diff
@@ -457,13 +456,18 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     ${TORCH_SRC_DIR}/csrc/jit/fuser/codegen.cpp
     ${TORCH_SRC_DIR}/csrc/jit/fuser/fallback.cpp
     ${TORCH_SRC_DIR}/csrc/jit/function.cpp
-    ${TORCH_ROOT}/test/cpp/jit/test.cpp
     )
 
+  if (NOT INTERN_BUILD_MOBILE)
+    list(APPEND TORCH_SRCS ${TORCH_SRC_DIR}/csrc/jit/export.cpp)
+    list(APPEND TORCH_SRCS ${TORCH_ROOT}/test/cpp/jit/test.cpp)
+    if(NOT WIN32)
```
Comment (ljk53) [inline caffe2/CMakeLists.txt]: nit: if (NOT WIN32)

File: caffe2/CMakeLists.txt [hunk 4]
```diff
@@ -497,7 +501,9 @@ if (NOT INTERN_BUILD_MOBILE OR NOT BUILD_CAFFE2_MOBILE)
     install(TARGETS caffe2_nvrtc DESTINATION "${TORCH_INSTALL_LIB_DIR}")
   endif()
 
-  if (NOT NO_API)
+  if (INTERN_BUILD_MOBILE)
```
Comment (ljk53) [inline caffe2/CMakeLists.txt]: do we need both INTERN_BUILD_MOBILE and NO_API here? can we use one and drop the other? e.g.:
option 1: keep NO_API and set it automatically in master CMakeLists.txt for mobile build; use if (NOT NO_API) ... else() ...;
option 2: drop NO_API and use if (INTERN_BUILD_MOBILE) ... else () ...;

"csrc/api/src/jit.cpp" is already listed on line 432, so we should be able to directly set NO_API in master CMakeLists for mobile build.

Comment (ljk53) [reply caffe2/CMakeLists.txt]: And can we make NO_API change a separate PR our rename this PR to better reflect your change.

You can use ghstack to maintain a stack of local diffs / PRs.

Comment (xta0) [reply caffe2/CMakeLists.txt]: There is one file - `output-archive.cpp` that depends on `export.h` is included via this `NO_API` flag. 


> And can we make NO_API change a separate PR our rename this PR to better reflect your change.

Yea, let's open another PR for this `NO_API` change

> do we need both INTERN_BUILD_MOBILE and NO_API here? can we use one and drop the other? e.g.:
> option 1: keep NO_API and set it automatically in master CMakeLists.txt for mobile build; use if (NOT NO_API) ... else() ...;
> option 2: drop NO_API and use if (INTERN_BUILD_MOBILE) ... else () ...;
> 
> "csrc/api/src/jit.cpp" is already listed on line 432, so we should be able to directly set NO_API in master CMakeLists for mobile build.

Actually I'm thinking if we can get rid of this `jit.cpp`, cuz there is only one function in the file, which does the compilation for cpp code. I've searched this symbol in the codebase, looks like we can safely remove it.

**Review by dzhulgakov — APPROVED** [review]

**Comment by xsacha** [comment]
Is this PR related to the latest Windows nightlies no longer having ONNX?

**Comment by dzhulgakov** [comment]
@xsacha - what exactly are you referring to by "no longer having ONNX"? from code it doesn't seem like it'd affect windows, but might make sense to investigate

**Comment by xsacha** [comment]
Latest nightlies in the past few days at least are missing it.

**Comment by bddppq** [comment]
@xsacha Which onnx files are you referring to?

**Pull request closed without merging**